### PR TITLE
Rich text: keep block ID on split

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This change ensures the block ID is kept for one of the split blocks, which is usually the first value unless it is empty and the second value is non-empty.
This also fixes an issue with the heading block where two paragraphs would be created if you split an empty heading in two. Instead, it should be a heading and a paragraph.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by talldan** [comment]
I think this should also fix https://github.com/WordPress/gutenberg/issues/26252.

File: packages/block-editor/src/components/rich-text/index.js
Comment (dratwas) [inline packages/block-editor/src/components/rich-text/index.js:301]: Hmmmm I'm not sure if I'm right but that means if we create a new block by pressing "return" we actually add a new block before the original one which is kinda confusing.

On the other hand, I'm wondering if it solves the issue from [here ](https://github.com/WordPress/gutenberg/issues/26252#issuecomment-716616579) 
 especially this point:
 
 - If there was anything attached to that entities' id, like metadata, that won't be there anymore. Or if anything references those ids, that tracking will be lost.
 
 
I think that the `clientID` should stay the same only for the true original block (the one that is selected when we hit the "return" key)

**Review by dratwas — COMMENTED** [review]
Thanks for this PR @ellatrix. Seems like it is almost the same as my PR :)

I'm still wondering what you meant in this comment by `Adding keepId is a horrible API.` https://github.com/WordPress/gutenberg/pull/28101/files#r564409051

What is the difference between your approach and mine? Is it only about the naming? (except the mutation of  a newly created block)

**Comment by ellatrix** [comment]
> What is the difference between your approach and mine? Is it only about the naming? (except the mutation of a newly created block)

* If after split the first block is empty and the second is not, the id should be kept for the second block.
* It also fixes a related issue with heading where it would replace a heading with two paragraphs if it is empty on split.

File: packages/block-editor/src/components/rich-text/index.js
Comment (ellatrix) [reply packages/block-editor/src/components/rich-text/index.js:301]: I think that makes sense. Nothing changes about the second block in that case and you insert a block before the selected block essentially. Think about a heading. If you press enter before the heading, you expect the heading to stay intact and a paragraph before the heading.

Comment (dratwas) [reply packages/block-editor/src/components/rich-text/index.js:301]: Just checked how it works in google docs and if you press enter at the beginning of the heading, the heading style is still there but I'm not a UX guy tho. However, I would love to hear more opinions about that.

**Comment by dratwas** [comment]
> - If after split the first block is empty and the second is not, the id should be kept for the second block.

I'm not a big fan of this approach and as mentioned it's not intuitive for me, however, if it makes sense for you I will not argue with that.

> - It also fixes a related issue with heading where it would replace a heading with two paragraphs if it is empty on split.

I haven't fixed this intentionally. I asked on the slack channel if it should work like that and was waiting for the response.  I also prefer to make one fixe per PR, especially in a project of a scale like this;)

Ok, closing my PR and waiting for the merge of this one since we still will need to add `InteractionManager` fix for mobile. However, I feel a bit confused since I think that this PR I almost the same as mine, and you could leave comments and discuss potential improvements there instead of creating a separate one :)

**Review by gwwar — APPROVED** [review]
This tests well for me @ellatrix! I think folks depending on stable block ids is a little unusual, and worth enforcing in unit tests (in a possible follow up) to avoid breakages and help document expected behavior. Some example cases might look like: press enter in middle of paragraph, start of paragraph, end of paragraph, etc.

Let's make sure the admin tests are green before merging, looks like we had a few timeouts.

@dratwas as far as I can tell the differences in the approach here, are mainly about what level this should occur at (it looks like we prefer this at a lower level) and the abstraction level, eg should a block be aware of an implementation detail like block id, vs telling rich text if it's the original block or not.

cc @TimothyBJacobs in case you wanted to take this for a spin.

**Pull request merged**
